// nexmo-cli/lib/client.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = void 0;

var _nexmo = _interopRequireDefault(require("nexmo"));

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

function _classCallCheck(instance, Constructor) { if (!(instance instanceof Constructor)) { throw new TypeError("Cannot call a class as a function"); } }

function _defineProperties(target, props) { for (var i = 0; i < props.length; i++) { var descriptor = props[i]; descriptor.enumerable = descriptor.enumerable || false; descriptor.configurable = true; if ("value" in descriptor) descriptor.writable = true; Object.defineProperty(target, descriptor.key, descriptor); } }

function _createClass(Constructor, protoProps, staticProps) { if (protoProps) _defineProperties(Constructor.prototype, protoProps); if (staticProps) _defineProperties(Constructor, staticProps); return Constructor; }

var Client =
/*#__PURE__*/
function () {
  function Client(config, emitter) {
    _classCallCheck(this, Client);

    this.emitter = emitter;
    this.config = config;
  }

  _createClass(Client, [{
    key: "instance",
    value: function instance() {
      return initialize(this.config, this.emitter);
    }
  }, {
    key: "definition",
    value: function definition() {
      return _nexmo.default;
    }
  }, {
    key: "instanceWith",
    value: function instanceWith(key, secret) {
      return new _nexmo.default({
        apiKey: key,
        apiSecret: secret
      });
    }
  }]);

  return Client;
}();

var _default = Client; // private methods

exports.default = _default;

var initialize = function initialize(config, emitter) {
  var packageDetails = require("".concat(__dirname, "/../package.json"));

  try {
    var credentials = config.read().credentials;
    return new _nexmo.default({
      apiKey: credentials.api_key,
      apiSecret: credentials.api_secret
    }, {
      debug: emitter.debugging,
      appendToUserAgent: "nexmo-cli/".concat(packageDetails.version.replace('v', ''))
    });
  } catch (e) {
    if (e instanceof TypeError) {
      emitter.error("Could not initialize Nexmo library. Please run 'nexmo setup <api_key> <api_secret>' to setup the CLI correctly. (".concat(e.message, ")"));
    } else {
      emitter.error("Could not read credentials. Please run 'nexmo setup <api_key> <api_secret>' to setup the CLI. (".concat(e.message, ")"));
    }
  }

  return null;
};
