// nexmo-cli/lib/config.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = void 0;

var _fs = _interopRequireDefault(require("fs"));

var _ini = _interopRequireDefault(require("ini"));

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

function _classCallCheck(instance, Constructor) { if (!(instance instanceof Constructor)) { throw new TypeError("Cannot call a class as a function"); } }

function _defineProperties(target, props) { for (var i = 0; i < props.length; i++) { var descriptor = props[i]; descriptor.enumerable = descriptor.enumerable || false; descriptor.configurable = true; if ("value" in descriptor) descriptor.writable = true; Object.defineProperty(target, descriptor.key, descriptor); } }

function _createClass(Constructor, protoProps, staticProps) { if (protoProps) _defineProperties(Constructor.prototype, protoProps); if (staticProps) _defineProperties(Constructor, staticProps); return Constructor; }

var Config =
/*#__PURE__*/
function () {
  function Config(emitter) {
    _classCallCheck(this, Config);

    this.emitter = emitter;
  }

  _createClass(Config, [{
    key: "read",
    value: function read() {
      this.credentials = _ini.default.parse(_fs.default.readFileSync(this.readFilename(), 'utf-8'));
      return this.credentials;
    }
  }, {
    key: "write",
    value: function write(data) {
      var local = arguments.length > 1 && arguments[1] !== undefined ? arguments[1] : false;

      _fs.default.writeFileSync(this.writeFilename(local), _ini.default.stringify(data));
    }
  }, {
    key: "readFilename",
    value: function readFilename() {
      var filename = localFile();

      if (_fs.default.existsSync(filename)) {
        return filename;
      }

      return homeFile();
    }
  }, {
    key: "writeFilename",
    value: function writeFilename() {
      var local = arguments.length > 0 && arguments[0] !== undefined ? arguments[0] : false;

      if (local) {
        return localFile();
      }

      return homeFile();
    }
  }, {
    key: "putAndSave",
    value: function putAndSave(values) {
      var writeLocal = arguments.length > 1 && arguments[1] !== undefined ? arguments[1] : false;
      var data = {};

      try {
        data = this.read();
      } catch (e) {
        this.emitter.warn('No existing config found. Writing to new file.');
      }

      for (var group in values) {
        data[group] = data[group] || {};
        var vals = values[group];

        for (var key in vals) {
          data[group][key] = vals[key];
        }
      }

      try {
        this.write(data, writeLocal);
        this.emitter.log("Credentials written to ".concat(this.writeFilename(writeLocal)));
      } catch (e) {
        this.emitter.error("Could not write credentials to ".concat(this.writeFilename(writeLocal), ". (").concat(e.message, ")"));
      }

      return data;
    }
  }]);

  return Config;
}();

var _default = Config; // private methods

exports.default = _default;

var localFile = function localFile() {
  return "".concat(process.cwd(), "/.nexmorc");
};

var homeFile = function homeFile() {
  var key = process.platform == 'win32' ? 'USERPROFILE' : 'HOME';
  return "".concat(process.env[key], "/.nexmorc");
};
